browser/brave_wallet/brave_wallet_provider_delegate_impl.cc

```diff
@@ -4,19 +4,90 @@
+#include "brave/components/brave_wallet/browser/brave_wallet_service.h"
+#include "brave/components/brave_wallet/browser/ethereum_permission_utils.h"
+#include "brave/components/brave_wallet/browser/hd_keyring.h"
+#include "brave/components/brave_wallet/browser/keyring_controller.h"
 #include "chrome/browser/ui/browser_finder.h"
+#include "components/content_settings/core/common/content_settings.h"
+#include "components/content_settings/core/common/content_settings_types.h"
+#include "components/permissions/permission_manager.h"
+#include "components/permissions/permissions_client.h"
+#include "content/public/browser/render_frame_host.h"
 #include "content/public/browser/web_contents.h"
 
 namespace brave_wallet {
 
+namespace {
+
+void OnRequestEthereumPermissions(
+    BraveWalletProviderDelegate::RequestEthereumPermissionsCallback callback,
+    const std::vector<ContentSetting>& responses) {
+  bool success = false;
+  for (auto response : responses) {
+    if (response == CONTENT_SETTING_ALLOW) {
+      success = true;
+      break;
+    }
+  }
+
+  std::move(callback).Run(success);
+}
+
+}  // namespace
+
 BraveWalletProviderDelegateImpl::BraveWalletProviderDelegateImpl(
-    content::WebContents* web_contents)
-    : web_contents_(web_contents) {}
+    content::WebContents* web_contents,
+    content::RenderFrameHost* const render_frame_host)
+    : web_contents_(web_contents),
+      routing_id_(render_frame_host->GetGlobalFrameRoutingId()) {}
+
+void BraveWalletProviderDelegateImpl::RequestEthereumPermissions(
+    RequestEthereumPermissionsCallback callback) {
+  permissions::PermissionManager* permission_manager =
+      permissions::PermissionsClient::Get()->GetPermissionManager(
+          web_contents_->GetBrowserContext());
+  if (!permission_manager) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  auto* service = BraveWalletServiceFactory::GetInstance()->GetForContext(
+      web_contents_->GetBrowserContext());
+  auto* keyring_controller = service->keyring_controller();
+  auto* default_keyring = keyring_controller->GetDefaultKeyring();
+  std::vector<std::string> addresses;
+  if (default_keyring) {
+    addresses = default_keyring->GetAccounts();
+  }
 
-void BraveWalletProviderDelegateImpl::ShowConnectToSiteUI() {
-  Browser* browser = chrome::FindBrowserWithWebContents(web_contents_);
-  brave::ShowWalletBubble(browser);
+  // To support ethereum permission to be per Ethereum account per site, we map
+  // each account address to one ethereum permission request. And the requests
+  // will have different origins which includes the address information. Here
+  // we first get a concatenated origin to include information of all wallet
+  // addresses, then adjust the origin of each request later in the process
+  // because PermissionManager::RequestPermissions only accepts a single origin
+  // parameter to be passes in.
+  GURL origin;
+  auto* rfh = content::RenderFrameHost::FromID(routing_id_);
```

Comment:
why are we using the routing_id_ here instead of just keeping a reference to the `rfh`? This class should not outlive the rfh, but even if it did, we're not checking for null `rfh` from the `routing_id_` so I'm curious why we're using it in the first place?

Reply:
rfh is checked now in `BraveEthereumPermissionContext::RequestPermissions` bfed18d